File: hc/packages/core/src/frame.ts
import { create, fromBinary, toBinary } from '@bufbuild/protobuf';
import { importP256VerifyingKey, signP1363LowS, verifyP1363LowS } from './crypto';
import {
  Direction,
  EncryptedFrameSchema,
  FrameType,
  WirePacketSchema,
} from './generated/mss/awp/v1/wire_pb';
import type { EndpointIdentity, P256PublicJwk } from './identity';
import type { OpenedSessionKeyPackage } from './session-key-package';

const textEncoder = new TextEncoder();

interface FrameBinding {
  readonly abaEndpointId: string;
  readonly hcEndpointId: string;
  readonly sessionId: string;
}

export async function createHCToABAFramePacket(
  identity: EndpointIdentity,
  keys: OpenedSessionKeyPackage,
  binding: FrameBinding,
  plaintext: Uint8Array,
  sequence: bigint,
  now = new Date(),
): Promise<Uint8Array> {
  const sessionId = decodeHexId(binding.sessionId);
  const abaEndpointId = decodeHexId(binding.abaEndpointId);
  const hcEndpointId = decodeHexId(binding.hcEndpointId);
  const messageId = crypto.getRandomValues(new Uint8Array(16));
  const channelId = await sessionChannelId(sessionId, abaEndpointId, hcEndpointId);
  const ciphertextLength = plaintext.length + 16;
  const aad = buildFrameAAD({
    channelId,
    ciphertextLength,
    createdAtMs: BigInt(now.getTime()),
    direction: Direction.HC_TO_ABA,
    flags: 0,
    keyGeneration: 1n,
    keyId: keys.material.keyId,
    messageId,
    receiverEndpointId: abaEndpointId,
    senderEndpointId: hcEndpointId,
    sequence,
    sessionId,
  });
  const ciphertext = await aesGcm('encrypt', keys.hcToAbaKey, keys.material.hcToAbaNoncePrefix, sequence, plaintext, aad);
  if (ciphertext.length !== ciphertextLength) {
    throw new Error('encrypted frame length is invalid');
  }
  const signature = await signP1363LowS(
    identity.signing.privateKey,
    frameSignatureInput(aad, ciphertext),
  );
  return toBinary(WirePacketSchema, create(WirePacketSchema, {
    body: {
      case: 'encrypted',
      value: create(EncryptedFrameSchema, {
        channelId,
        ciphertext,
        createdAtMs: BigInt(now.getTime()),
        cryptoSuiteId: 1,
        direction: Direction.HC_TO_ABA,
        flags: 0,
        frameType: FrameType.ACP_TRANSPORT_FRAME,
        keyGeneration: 1n,
        keyId: keys.material.keyId,
        messageId,
        receiverEndpointId: abaEndpointId,
        senderEndpointId: hcEndpointId,
        sequence,
        sessionId,
        signature,
      }),
    },
    packetId: crypto.getRandomValues(new Uint8Array(16)),
    wireMajor: 1,
    wireMinor: 0,
  }));
}

export interface OpenedFrame {
  readonly plaintext: Uint8Array;
  readonly sequence: bigint;
}

export async function openABAToHCFramePacket(
  abaSigningPublicJwk: P256PublicJwk,
  keys: OpenedSessionKeyPackage,
  binding: FrameBinding,
  encoded: Uint8Array,
): Promise<OpenedFrame | null> {
  const packet = fromBinary(WirePacketSchema, encoded);
  if (packet.body.case !== 'encrypted') {
    return null;
  }
  if (packet.wireMajor !== 1 || packet.wireMinor !== 0 || packet.packetId.length !== 16) {
    throw new Error('encrypted AWP packet is invalid');
  }
  const frame = packet.body.value;
  const sessionId = decodeHexId(binding.sessionId);
  const abaEndpointId = decodeHexId(binding.abaEndpointId);
  const hcEndpointId = decodeHexId(binding.hcEndpointId);
  const channelId = await sessionChannelId(sessionId, abaEndpointId, hcEndpointId);
  if (
    frame.cryptoSuiteId !== 1 ||
    frame.frameType !== FrameType.ACP_TRANSPORT_FRAME ||
    frame.flags !== 0 ||
    frame.messageId.length !== 16 ||
    !equalBytes(frame.channelId, channelId) ||
    !equalBytes(frame.sessionId, sessionId) ||
    !equalBytes(frame.senderEndpointId, abaEndpointId) ||
    !equalBytes(frame.receiverEndpointId, hcEndpointId) ||
    frame.direction !== Direction.ABA_TO_HC ||
    frame.sequence <= 0n ||
    frame.keyGeneration !== 1n ||
    !equalBytes(frame.keyId, keys.material.keyId) ||
    frame.ciphertext.length < 16 ||
    frame.ciphertext.length > 1_048_576 ||
    frame.signature.length !== 64
  ) {
    throw new Error('encrypted frame binding is invalid');
  }
  const aad = buildFrameAAD({
    channelId: frame.channelId,
    ciphertextLength: frame.ciphertext.length,
    createdAtMs: frame.createdAtMs,
    direction: frame.direction,
    flags: frame.flags,
    keyGeneration: frame.keyGeneration,
    keyId: frame.keyId,
    messageId: frame.messageId,
    receiverEndpointId: frame.receiverEndpointId,
    senderEndpointId: frame.senderEndpointId,
    sequence: frame.sequence,
    sessionId: frame.sessionId,
  });
  const signingKey = await importP256VerifyingKey(abaSigningPublicJwk);
  if (!(await verifyP1363LowS(signingKey, frameSignatureInput(aad, frame.ciphertext), frame.signature))) {
    throw new Error('encrypted frame signature is invalid');
  }
  const plaintext = await aesGcm(
    'decrypt', keys.abaToHcKey, keys.material.abaToHcNoncePrefix,
    frame.sequence, frame.ciphertext, aad,
  );
  const text = new TextDecoder('utf-8', { fatal: true }).decode(plaintext);
  JSON.parse(text) as unknown;
  return { plaintext, sequence: frame.sequence };
}

interface FrameAADInput {
  readonly channelId: Uint8Array;
  readonly ciphertextLength: number;
  readonly createdAtMs: bigint;
  readonly direction: Direction;
  readonly flags: number;
  readonly keyGeneration: bigint;
  readonly keyId: Uint8Array;
  readonly messageId: Uint8Array;
  readonly receiverEndpointId: Uint8Array;
  readonly senderEndpointId: Uint8Array;
  readonly sequence: bigint;
  readonly sessionId: Uint8Array;
}

export function buildFrameAAD(input: FrameAADInput): Uint8Array {
  for (const value of [
    input.messageId, input.channelId, input.sessionId,
    input.senderEndpointId, input.receiverEndpointId, input.keyId,
  ]) {
    if (value.length !== 16 || value.every((byte) => byte === 0)) {
      throw new Error('frame AAD identifier is invalid');
    }
  }
  if (
    input.flags < 0 || input.flags > 0xffff ||
    input.sequence <= 0n || input.keyGeneration <= 0n ||
    input.createdAtMs < 0n || input.ciphertextLength < 16 || input.ciphertextLength > 1_048_576 ||
    ![Direction.HC_TO_ABA, Direction.ABA_TO_HC].includes(input.direction)
  ) {
    throw new Error('frame AAD field is invalid');
  }
  const output = new Uint8Array(148);
  output.set(textEncoder.encode('AWP1'), 0);
  const view = new DataView(output.buffer);
  view.setUint16(4, 1, false);
  view.setUint16(6, 0, false);
  view.setUint16(8, 1, false);
  view.setUint16(10, 1, false);
  view.setUint32(12, input.flags, false);
  output.set(input.messageId, 16);
  output.set(input.channelId, 32);
  output.set(input.sessionId, 48);
  output.set(input.senderEndpointId, 64);
  output.set(input.receiverEndpointId, 80);
  output[96] = input.direction;
  view.setBigUint64(104, input.sequence, false);
  view.setBigUint64(112, input.keyGeneration, false);
  output.set(input.keyId, 120);
  view.setBigInt64(136, input.createdAtMs, false);
  view.setUint32(144, input.ciphertextLength, false);
  return output;
}

export async function sessionChannelId(
  sessionId: Uint8Array,
  abaEndpointId: Uint8Array,
  hcEndpointId: Uint8Array,
): Promise<Uint8Array> {
  const digest = await crypto.subtle.digest('SHA-256', asArrayBuffer(concatenate([
    textEncoder.encode('mss-awp-channel-v1'), sessionId, abaEndpointId, hcEndpointId,
  ])));
  return new Uint8Array(digest).slice(0, 16);
}

async function aesGcm(
  operation: 'encrypt' | 'decrypt',
  key: Uint8Array,
  prefix: Uint8Array,
  sequence: bigint,
  input: Uint8Array,
  aad: Uint8Array,
): Promise<Uint8Array> {
  if (key.length !== 32 || prefix.length !== 4 || sequence <= 0n) {
    throw new Error('frame key or nonce is invalid');
  }
  const cryptoKey = await crypto.subtle.importKey('raw', asArrayBuffer(key), 'AES-GCM', false, [operation]);
  const nonce = new Uint8Array(12);
  nonce.set(prefix);
  new DataView(nonce.buffer).setBigUint64(4, sequence, false);
  try {
    const params = { additionalData: asArrayBuffer(aad), iv: nonce.buffer, name: 'AES-GCM', tagLength: 128 };
    const result = operation === 'encrypt'
      ? await crypto.subtle.encrypt(params, cryptoKey, asArrayBuffer(input))
      : await crypto.subtle.decrypt(params, cryptoKey, asArrayBuffer(input));
    return new Uint8Array(result);
  } catch {
    throw new Error('frame AEAD operation failed');
  }
}

function frameSignatureInput(aad: Uint8Array, ciphertext: Uint8Array): Uint8Array {
  return concatenate([textEncoder.encode('mss-awp-frame-signature-v1'), aad, ciphertext]);
}

function decodeHexId(value: string): Uint8Array {
  if (!/^[0-9a-f]{32}$/u.test(value)) {
    throw new Error('frame identifier is invalid');
  }
  return Uint8Array.from({ length: 16 }, (_, index) => Number.parseInt(value.slice(index * 2, index * 2 + 2), 16));
}

function concatenate(values: readonly Uint8Array[]): Uint8Array {
  const result = new Uint8Array(values.reduce((sum, value) => sum + value.length, 0));
  let offset = 0;
  for (const value of values) {
    result.set(value, offset);
    offset += value.length;
  }
  return result;
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function asArrayBuffer(value: Uint8Array): ArrayBuffer {
  const copy = new Uint8Array(value.byteLength);
  copy.set(value);
  return copy.buffer;
}

File: aba/src/crypto/frame.rs
use aes_gcm::{
    Aes256Gcm,
    aead::{Aead as _, KeyInit as _, Payload, array::Array},
};
use p256::ecdsa::{SigningKey, VerifyingKey};
use sha2::{Digest as _, Sha256};
use thiserror::Error;
use zeroize::Zeroizing;

use super::{sign_p1363_low_s, verify_p1363_low_s};
use crate::wire::{FrameAadV1, WireValidationError};

pub struct ProtectedFrame {
    pub aad: [u8; 148],
    pub ciphertext: Vec<u8>,
    pub signature: [u8; 64],
}

#[derive(Debug, Error)]
pub enum FrameCryptoError {
    #[error("frame input is invalid")]
    InvalidInput,
    #[error("frame AAD is invalid")]
    Aad(#[from] WireValidationError),
    #[error("frame signature is invalid")]
    Signature,
    #[error("frame AEAD operation failed")]
    Aead,
}

pub fn seal_frame(
    key: &[u8; 32],
    nonce_prefix: &[u8; 4],
    signing_key: &SigningKey,
    mut aad: FrameAadV1,
    plaintext: &[u8],
) -> Result<ProtectedFrame, FrameCryptoError> {
    if plaintext.is_empty() || plaintext.len() > (1 << 20) - 16 {
        return Err(FrameCryptoError::InvalidInput);
    }
    aad.ciphertext_length =
        u32::try_from(plaintext.len() + 16).map_err(|_| FrameCryptoError::InvalidInput)?;
    let encoded_aad = aad.encode()?;
    let nonce = frame_nonce(nonce_prefix, aad.sequence)?;
    let cipher = Aes256Gcm::new(&Array(*key));
    let ciphertext = cipher
        .encrypt(
            &Array(nonce),
            Payload {
                msg: plaintext,
                aad: &encoded_aad,
            },
        )
        .map_err(|_| FrameCryptoError::Aead)?;
    let signature = sign_p1363_low_s(
        signing_key,
        &frame_signature_input(&encoded_aad, &ciphertext),
    );
    Ok(ProtectedFrame {
        aad: encoded_aad,
        ciphertext,
        signature,
    })
}

pub fn open_frame(
    key: &[u8; 32],
    nonce_prefix: &[u8; 4],
    verifying_key: &VerifyingKey,
    aad: &FrameAadV1,
    ciphertext: &[u8],
    signature: &[u8],
) -> Result<Zeroizing<Vec<u8>>, FrameCryptoError> {
    let encoded_aad = aad.encode()?;
    if usize::try_from(aad.ciphertext_length).ok() != Some(ciphertext.len())
        || ciphertext.len() < 16
    {
        return Err(FrameCryptoError::InvalidInput);
    }
    if !verify_p1363_low_s(
        verifying_key,
        &frame_signature_input(&encoded_aad, ciphertext),
        signature,
    ) {
        return Err(FrameCryptoError::Signature);
    }
    let nonce = frame_nonce(nonce_prefix, aad.sequence)?;
    let cipher = Aes256Gcm::new(&Array(*key));
    let plaintext = cipher
        .decrypt(
            &Array(nonce),
            Payload {
                msg: ciphertext,
                aad: &encoded_aad,
            },
        )
        .map_err(|_| FrameCryptoError::Aead)?;
    Ok(Zeroizing::new(plaintext))
}

pub fn frame_nonce(prefix: &[u8; 4], sequence: u64) -> Result<[u8; 12], FrameCryptoError> {
    if sequence == 0 {
        return Err(FrameCryptoError::InvalidInput);
    }
    let mut nonce = [0_u8; 12];
    nonce[..4].copy_from_slice(prefix);
    nonce[4..].copy_from_slice(&sequence.to_be_bytes());
    Ok(nonce)
}

pub fn session_channel_id(
    session_id: &[u8; 16],
    aba_endpoint_id: &[u8; 16],
    hc_endpoint_id: &[u8; 16],
) -> Result<[u8; 16], FrameCryptoError> {
    if session_id.iter().all(|value| *value == 0)
        || aba_endpoint_id.iter().all(|value| *value == 0)
        || hc_endpoint_id.iter().all(|value| *value == 0)
        || aba_endpoint_id == hc_endpoint_id
    {
        return Err(FrameCryptoError::InvalidInput);
    }
    let mut digest = Sha256::new();
    digest.update(b"mss-awp-channel-v1");
    digest.update(session_id);
    digest.update(aba_endpoint_id);
    digest.update(hc_endpoint_id);
    let result = digest.finalize();
    let mut id = [0_u8; 16];
    id.copy_from_slice(&result[..16]);
    Ok(id)
}

pub fn frame_content_hash(aad: &[u8; 148], ciphertext: &[u8], signature: &[u8]) -> [u8; 32] {
    let mut digest = Sha256::new();
    digest.update(aad);
    digest.update(ciphertext);
    digest.update(signature);
    digest.finalize().into()
}

fn frame_signature_input(aad: &[u8; 148], ciphertext: &[u8]) -> Vec<u8> {
    let mut input = Vec::with_capacity(26 + aad.len() + ciphertext.len());
    input.extend_from_slice(b"mss-awp-frame-signature-v1");
    input.extend_from_slice(aad);
    input.extend_from_slice(ciphertext);
    input
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::wire::{CryptoSuite, Direction, FrameType};

    #[test]
    fn seals_verifies_and_opens_frame() -> Result<(), Box<dyn std::error::Error>> {
        let signing = SigningKey::from_slice(&[9_u8; 32])?;
        let aad = FrameAadV1 {
            crypto_suite: CryptoSuite::Suite0001,
            frame_type: FrameType::AcpTransportFrame,
            flags: 0,
            message_id: [1_u8; 16],
            channel_id: session_channel_id(&[2_u8; 16], &[3_u8; 16], &[4_u8; 16])?,
            session_id: [2_u8; 16],
            sender_endpoint_id: [4_u8; 16],
            receiver_endpoint_id: [3_u8; 16],
            direction: Direction::HcToAba,
            sequence: 1,
            key_generation: 1,
            key_id: [5_u8; 16],
            created_at_ms: 1_800_000_000_000,
            ciphertext_length: 0,
        };
        let plaintext =
            br#"{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":1}}"#;
        let protected = seal_frame(&[6_u8; 32], &[7_u8; 4], &signing, aad.clone(), plaintext)?;
        let mut opened_aad = aad;
        opened_aad.ciphertext_length = u32::try_from(protected.ciphertext.len())?;
        assert_eq!(
            open_frame(
                &[6_u8; 32],
                &[7_u8; 4],
                signing.verifying_key(),
                &opened_aad,
                &protected.ciphertext,
                &protected.signature,
            )?
            .as_slice(),
            plaintext
        );
        let mut tampered = protected.ciphertext.clone();
        tampered[0] ^= 1;
        assert!(matches!(
            open_frame(
                &[6_u8; 32],
                &[7_u8; 4],
                signing.verifying_key(),
                &opened_aad,
                &tampered,
                &protected.signature,
            ),
            Err(FrameCryptoError::Signature)
        ));
        Ok(())
    }
}
